File: kmp/monero-core/src/commonMain/kotlin/io/monero/core/Mnemonic.kt
package io.monero.core

import io.monero.crypto.Keccak
import io.monero.crypto.Ed25519
import kotlinx.serialization.Serializable

/**
 * Monero mnemonic seed operations.
 *
 * Monero uses 25-word mnemonic seeds based on the Electrum word list.
 * The seed encodes a 256-bit private spend key + 1 checksum word.
 *
 * Word list languages supported:
 * - English (1626 words)
 * - Japanese, Spanish, Portuguese, French, German, Italian, Russian, etc.
 */
object Mnemonic {
    private const val SEED_LENGTH = 32
    private const val WORD_COUNT = 25
    private const val CHECKSUM_INDEX = 24
    private const val WORDS_PER_CHUNK = 3
    private const val WORD_LIST_SIZE = 1626

    /**
     * English word list (first few words for structure - full list in resources)
     */
    private val ENGLISH_WORD_LIST = listOf(
        "abbey", "abducts", "ability", "ablaze", "abnormal", "abort", "abrasive", "absorb",
        // ... 1626 words total - loaded from resources in production
    )

    /**
     * Convert entropy to mnemonic words
     *
     * @param entropy 32-byte random data
     * @param language Word list language (default: English)
     * @return List of 25 mnemonic words
     */
    fun entropyToMnemonic(entropy: ByteArray, language: String = "english"): List<String> {
        require(entropy.size == SEED_LENGTH) { "Entropy must be $SEED_LENGTH bytes" }

        val wordList = getWordList(language)
        val words = mutableListOf<String>()

        // Convert entropy to words (3 words per 4 bytes chunk)
        for (i in 0 until 8) {
            val chunk = (entropy[i * 4].toUInt() and 0xFFu) or
                    ((entropy[i * 4 + 1].toUInt() and 0xFFu) shl 8) or
                    ((entropy[i * 4 + 2].toUInt() and 0xFFu) shl 16) or
                    ((entropy[i * 4 + 3].toUInt() and 0xFFu) shl 24)

            val w1 = (chunk % WORD_LIST_SIZE.toUInt()).toInt()
            val w2 = (((chunk / WORD_LIST_SIZE.toUInt()) + w1.toUInt()) % WORD_LIST_SIZE.toUInt()).toInt()
            val w3 = (((chunk / WORD_LIST_SIZE.toUInt() / WORD_LIST_SIZE.toUInt()) + w2.toUInt()) % WORD_LIST_SIZE.toUInt()).toInt()

            words.add(wordList[w1])
            words.add(wordList[w2])
            words.add(wordList[w3])
        }

        // Add checksum word
        val checksumWord = computeChecksum(words, wordList)
        words.add(checksumWord)

        return words
    }

    /**
     * Convert mnemonic words to entropy
     *
     * @param words List of 25 mnemonic words
     * @param language Word list language
     * @return 32-byte entropy
     */
    fun mnemonicToEntropy(words: List<String>, language: String = "english"): ByteArray {
        require(words.size == WORD_COUNT) { "Mnemonic must have $WORD_COUNT words" }

        val wordList = getWordList(language)
        val seedWords = words.take(24)
        val checksumWord = words[24]

        // Verify checksum
        val expectedChecksum = computeChecksum(seedWords, wordList)
        require(checksumWord == expectedChecksum) { "Invalid checksum word" }

        // Convert words to entropy
        val entropy = ByteArray(SEED_LENGTH)
        for (i in 0 until 8) {
            val w1 = wordList.indexOf(seedWords[i * 3])
            val w2 = wordList.indexOf(seedWords[i * 3 + 1])
            val w3 = wordList.indexOf(seedWords[i * 3 + 2])

            require(w1 >= 0 && w2 >= 0 && w3 >= 0) { "Unknown word in mnemonic" }

            val chunk = (w1.toUInt() +
                    WORD_LIST_SIZE.toUInt() * ((WORD_LIST_SIZE.toUInt() + w2.toUInt() - w1.toUInt()) % WORD_LIST_SIZE.toUInt()) +
                    WORD_LIST_SIZE.toUInt() * WORD_LIST_SIZE.toUInt() * ((WORD_LIST_SIZE.toUInt() + w3.toUInt() - w2.toUInt()) % WORD_LIST_SIZE.toUInt()))

            entropy[i * 4] = (chunk and 0xFFu).toByte()
            entropy[i * 4 + 1] = ((chunk shr 8) and 0xFFu).toByte()
            entropy[i * 4 + 2] = ((chunk shr 16) and 0xFFu).toByte()
            entropy[i * 4 + 3] = ((chunk shr 24) and 0xFFu).toByte()
        }

        return entropy
    }

    /**
     * Validate mnemonic words
     */
    fun validate(words: List<String>, language: String = "english"): Boolean {
        if (words.size != WORD_COUNT) return false

        return try {
            mnemonicToEntropy(words, language)
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Generate a new random mnemonic
     */
    fun generate(language: String = "english"): List<String> {
        val entropy = generateSecureRandom(SEED_LENGTH)
        return entropyToMnemonic(entropy, language)
    }

    private fun computeChecksum(words: List<String>, wordList: List<String>): String {
        // Checksum is based on first 3 characters of each word
        val prefixString = words.joinToString("") {
            it.take(prefixLength(wordList))
        }
        val crc = crc32(prefixString.encodeToByteArray())
        return words[(crc % words.size.toUInt()).toInt()]
    }

    private fun prefixLength(wordList: List<String>): Int {
        // English uses 3-character prefix for unique identification
        return 3
    }

    private fun crc32(data: ByteArray): UInt {
        var crc = 0xFFFFFFFFu
        for (byte in data) {
            crc = crc xor byte.toUInt()
            for (j in 0 until 8) {
                crc = if (crc and 1u != 0u) {
                    (crc shr 1) xor 0xEDB88320u
                } else {
                    crc shr 1
                }
            }
        }
        return crc xor 0xFFFFFFFFu
    }

    private fun getWordList(language: String): List<String> {
        return when (language.lowercase()) {
            "english" -> ENGLISH_WORD_LIST
            else -> throw IllegalArgumentException("Unsupported language: $language")
        }
    }
}

/**
 * Generate cryptographically secure random bytes.
 * Platform-specific implementation required.
 */
expect fun generateSecureRandom(size: Int): ByteArray

File: dart/lib/src/core/mnemonic.dart
import 'dart:math';
import 'dart:typed_data';

import '../constants.dart';
import 'english_word_list.dart';

/// Monero mnemonic seed operations.
///
/// Monero uses 25-word mnemonic seeds based on the Electrum word list.
/// The seed encodes a 256-bit private spend key + 1 checksum word.
///
/// Word list languages supported:
/// - English (1626 words)
/// - Japanese, Spanish, Portuguese, French, German, Italian, Russian, etc.
class Mnemonic {
  static const int _seedLength = mnemonicSeedBytes;
  static const int _wordCount = mnemonicWordCount;
  static const int _wordListSize = 1626;

  /// Convert entropy to mnemonic words.
  ///
  /// [entropy] 32-byte random data
  /// [language] Word list language (default: English)
  /// Returns list of 25 mnemonic words
  static List<String> entropyToMnemonic(
    Uint8List entropy, {
    String language = 'english',
  }) {
    if (entropy.length != _seedLength) {
      throw ArgumentError('Entropy must be $_seedLength bytes');
    }

    final wordList = _getWordList(language);
    final words = <String>[];

    // Convert entropy to words (3 words per 4 bytes chunk)
    for (var i = 0; i < 8; i++) {
      final chunk = (entropy[i * 4] & 0xFF) |
          ((entropy[i * 4 + 1] & 0xFF) << 8) |
          ((entropy[i * 4 + 2] & 0xFF) << 16) |
          ((entropy[i * 4 + 3] & 0xFF) << 24);

      final w1 = chunk % _wordListSize;
      final w2 = ((chunk ~/ _wordListSize) + w1) % _wordListSize;
      final w3 =
          ((chunk ~/ _wordListSize ~/ _wordListSize) + w2) % _wordListSize;

      words
        ..add(wordList[w1])
        ..add(wordList[w2])
        ..add(wordList[w3]);
    }

    // Add checksum word
    final checksumWord = _computeChecksum(words, wordList);
    words.add(checksumWord);

    return words;
  }

  /// Convert mnemonic words to entropy.
  ///
  /// [words] List of 25 mnemonic words
  /// [language] Word list language
  /// Returns 32-byte entropy
  static Uint8List mnemonicToEntropy(
    List<String> words, {
    String language = 'english',
  }) {
    if (words.length != _wordCount) {
      throw ArgumentError('Mnemonic must have $_wordCount words');
    }

    final wordList = _getWordList(language);
    final seedWords = words.sublist(0, 24);
    final checksumWord = words[24];

    // Verify checksum
    final expectedChecksum = _computeChecksum(seedWords, wordList);
    if (checksumWord != expectedChecksum) {
      throw const FormatException('Invalid checksum word');
    }

    // Convert words to entropy
    final entropy = Uint8List(_seedLength);
    for (var i = 0; i < 8; i++) {
      final w1 = wordList.indexOf(seedWords[i * 3]);
      final w2 = wordList.indexOf(seedWords[i * 3 + 1]);
      final w3 = wordList.indexOf(seedWords[i * 3 + 2]);

      if (w1 < 0 || w2 < 0 || w3 < 0) {
        throw const FormatException('Unknown word in mnemonic');
      }

      final chunk = w1 +
          _wordListSize * ((_wordListSize + w2 - w1) % _wordListSize) +
          _wordListSize *
              _wordListSize *
              ((_wordListSize + w3 - w2) % _wordListSize);

      entropy[i * 4] = chunk & 0xFF;
      entropy[i * 4 + 1] = (chunk >> 8) & 0xFF;
      entropy[i * 4 + 2] = (chunk >> 16) & 0xFF;
      entropy[i * 4 + 3] = (chunk >> 24) & 0xFF;
    }

    return entropy;
  }

  /// Validate mnemonic words.
  static bool validate(List<String> words, {String language = 'english'}) {
    if (words.length != _wordCount) return false;

    try {
      mnemonicToEntropy(words, language: language);
      return true;
    } catch (_) {
      return false;
    }
  }

  /// Generate a new random mnemonic.
  static List<String> generate({String language = 'english'}) {
    final entropy = _generateSecureRandom(_seedLength);
    return entropyToMnemonic(entropy, language: language);
  }

  static String _computeChecksum(List<String> words, List<String> wordList) {
    // Checksum is based on first 3 characters of each word
    final prefixString = words.map((w) => w.substring(0, min(3, w.length))).join();
    final crc = _crc32(Uint8List.fromList(prefixString.codeUnits));
    return words[crc % words.length];
  }

  static int _crc32(Uint8List data) {
    var crc = 0xFFFFFFFF;
    for (final byte in data) {
      crc ^= byte;
      for (var j = 0; j < 8; j++) {
        if ((crc & 1) != 0) {
          crc = (crc >> 1) ^ 0xEDB88320;
        } else {
          crc >>= 1;
        }
      }
    }
    return crc ^ 0xFFFFFFFF;
  }

  static Uint8List _generateSecureRandom(int size) {
    final random = Random.secure();
    return Uint8List.fromList(
      List.generate(size, (_) => random.nextInt(256)),
    );
  }

  static List<String> _getWordList(String language) {
    return switch (language.toLowerCase()) {
      'english' => englishWordList,
      _ => throw ArgumentError('Unsupported language: $language'),
    };
  }
}
